### ISSUE TITLE
Questions about Writer, WriterT and creating `from___` methods

### ISSUE THREAD

**Issue by atomanyih** [gcanti/fp-ts#1710]
I've been working with Writer and WriterT to define some other monads (`TaskWriter`, `TaskWriterEither`, `ReaderTaskWriterEither`) and one thing I don't understand is why using `WriterT` on `Task`:

```ts
const writerM = WT.getWriterM(T.Monad);
const TaskWriterMonadTypeClass = writerM.getMonad<string[]>(A.getMonoid());
```
creates a monad structured like this:
```ts
type TaskWriter<A> = {
  (): T.Task<[A, string[]]>;
};
```

as opposed to
```ts
type TaskWriter<A> = T.Task<W.Writer<A, string[]>>;
```

I figure there's a good reason for this, but am finding it confusing especially as it seems to make it harder to write lifting methods like `TaskWriterEither.fromTaskEither`

```ts
// TaskWriterEither.ts
export const fromTaskEither: <E, A>(
  taskEither: TE.TaskEither<E, A>,
) => TaskWriterEither<E, A> = (taskEither) => () => () => {
  return taskEither().then((ea) => {
    return [ea, []];
  });
};
```

Maybe I'm missing a more elegant way of writing this, but it feels like we're breaking an abstraction and repeating work from `TaskWriter` when we manually `then`/`await` and provide the empty monoid value.

Apologies if this question is too open-ended, it's been difficult to wrap my brain around some of the abstraction

**Comment by waynevanson** [comment]
Writer is lazy, which is why it is in a thunk, same with Task.
Can't give you a good reason for double lazy besides the fact that Task is inside Writer. Writer is not `[A, W]`, it's `() => [A,W]`.

Writing those transformations should be done with signature `NaturalTransformation<taskEither.URI, URI>` for ease.

**Comment by waynevanson** [comment]
@atomanyih 
The reason it's a thunk is so the `W` is hidden and enforces that the Writer is either readonly, or modify in place.

This is the most useful article I've found regarding Writer.
https://mmhaskell.com/monads/reader-writer#using-writier-to-track-the-accumulator

**Comment by gcanti** [comment]
I agree with @atomanyih that the current encoding is convoluted (my mistake), the best encoding should be

```ts
export type Writer<E, A> = readonly [A, E]
```

i.e. `Writer` = `ReadonlyTuple`
